keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1249,6 +1256,181 @@ of `processing latency` used for determining the weight of the queue may
+  hash(ns, name) = 1 + A + hash(namespace/name)%B otherwise
+```
+For small enough A and B (e.g. A=3, B=6), the representation should have less
+than 1000 entries, so it would be small enough to make periodic updates in etcd
+reasonable.
+
+We can optimize amount of data written to etcd by frequently (say once per
+second) checking what changed, but writing rarely (say once per minute) or
+if values in some buckets significantly increased.
+The above algorithm would allow us to avoid some more complicated time-smearing,
+as whenever something quickly grows we report it, but we don't immediately
+downscale which is a way to somehow incorporate a history.
+
+However, we will treat the above as a feasibility proof.  We will just start
+with the simplest apprach of treating each kube-apiserver independently.
+We will implement the above (i.e. knowledge sharing between kube-apiserver),
+if the independence assumption will not work good enough.
+The above description shows that it won't result in almost any wasted work
+if the code will be well structured.
+
+##### Cost of the watch event
+
+We assumed above that the cost of processing every watch event is equal.
+However, in practice the cost associated with sending an event consists
+of two main parts:
+- the cost of going through event change logic
+- the cost of processing the even object (e.g. deserialization or sending data
+  over network)
+The first one is close to equal independently of the event, the second one
+is more proportional to the size of the object.
+However, estimating size of the is hard to predict for PATCH or DELETE requests.
+Additionally, even for POST or PUT requests where we could potentially
+estimate it based on size of the body of the requests, we may not yet have
+access to it when we need to make a decision.
+
+One way to estimate it would be to keep a running average of watch event size
+per bucket.  While it won't give us an accurate estimate, it should amortize
+well over time.
+
+Obviously some normalization will be needed here, but it's impossible to assess
+it on paper, and we are leaving this for the implementation and tuning phase
+to figure out the details.
+
+##### Dispatching the request
+
+We described how we can estimate the cost of the request associated with
+watch events triggerred by it.  But we didn't yet said how this translates
+to dispatching the request.
+
+First of all, we need to decide how to translate the estimated cost to the
+`width` of the request and its latency.  Second, we need to introduce the
+changes to our virtual world, as the fact that the request finished doesn't
+mean that sending all associated watch events has also finished (as they are
+send asynchronously).
+
+Given that individual watch events are to significant extent processed
+independently in individual goroutines, it actually makes sense to adjust the
+`width` of the request based on the expected number of triggerred events.
```

Comment:
Width of a given mutating request's watch events is really hard to predict because effectively it depends on all the events for the other changes, too.

I think I'm assuming apiserver is bandwidth-limited and has a queue of outgoing watch notifications at all times. I think you're assuming it's not bandwidth limited and therefore the events sent per unit time is directly related to the number of incoming changes.

It occurs to me that we could just look at the watch queues to see which condition we're in.

We could also keep running stats on the event egress rate. If we know we currently have a rate of R events/second, and a given mutating request is going to add approximately N events, then we can predict that it will take an additional N*R seconds to send all the events from this request. It's not obvious to me that we'd get any benefit from trying to estimate how many of these events will go out in parallel, and that definitely makes it a lot more complicated.

Reply:
> Width of a given mutating request's watch events is really hard to predict because effectively it depends on all the events for the other changes, too.

Are you talking about fan-out here (i.e. the upcoming mutating requests that will be triggered by this mutating request)? I'm conceptually ignoring it, because that will be captured by that other mutating request width.

> I think I'm assuming apiserver is bandwidth-limited and has a queue of outgoing watch notifications at all times. I think you're assuming it's not bandwidth limited and therefore the events sent per unit time is directly related to the number of incoming changes.

I'm also assuming it's bandwdith-limited (we've seen multiple times a network-starvation in our prod clusters).

And this is exactly why I'm saying that we should adjust width here too. `<width> x <latency>` translates to some network bandwith and that is better mimicing it imho.

> We could also keep running stats on the event egress rate. If we know we currently have a rate of R events/second, and a given mutating request is going to add approximately N events, then we can predict that it will take an additional N*R seconds to send all the events from this request. It's not obvious to me that we'd get any benefit from trying to estimate how many of these events will go out in parallel, and that definitely makes it a lot more complicated.

My point is that event processing is not independent from other requests.
How I conceptually think about that is that with a request with `width=<width>` that lasts for `<duration>`, I'm effectively blocking `<width>` (should be normalized) percentage of capacity for `<duration>`. 

In practice, given that sending watch events is highly parallelized, we don't have a way for enforcing that those events actually will be processed in small fraction of capacity and those will be starving other requests.

If you feel strongly, I'm fine with starting with `width=1` and only play with latency (the further changes will be triviail, because we need to implement generic width for lists anyway - so it will just translate to changes in a function that is estimating width) - but I'm 99% sure this won't get us where we would like to be.

Reply:
> And this is exactly why I'm saying that we should adjust width here too. <width> x <latency> translates to some network bandwith and that is better mimicing it imho.

Sure. I'm just claiming you don't know what fraction of the network bandwidth you're getting because you don't know how many other watch events are going on at the moment.

> I'm fine with starting with width=1 and only play with latency ...

I don't feel super strongly, and I think it's clear that 1 is *not* the correct value, I just don't know how to compute the correct value. I don't think the formula you give below would be disastrous or anything.

Reply:
> Sure. I'm just claiming you don't know what fraction of the network bandwidth you're getting because you don't know how many other watch events are going on at the moment.

The throttling that we get from using this mechanism should let us keep the "to-be-send watch event" under so, while we may see some short spikes, it should work fine, right?

> I don't feel super strongly, and I think it's clear that 1 is not the correct value, I just don't know how to compute the correct value. I don't think the formula you give below would be disastrous or anything.

I'm going to leave as is then - fortunately this is the part that is super easy to change later, so I don't think we need to be super accurate here.

Reply:
Yeah I think this is fine as a starting point, we can fix it later if it's not fine :)